### PR TITLE
Api integration

### PR DESCRIPTION
**Pull request by SsopoWolox**
Summary

Completed Creation of DTOs class
Completed Integration external Service
Completed Creation of method findByIsbn
Completed run module training
Completed versioning project for this version

Trello Card
https://trello.com/c/TGFlonWZ/22-integraci%C3%B3n-api-externa

### CODE REVIEW

File: src/main/java/wolox/training/integration/dto/Book.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    @JsonProperty("authors")
+    private List<Author> authors = null;
+    @JsonProperty("classifications")
+    private Classifications classifications;
+    @JsonProperty("publish_places")
+    private List<PublishPlace> publishPlaces = null;
+    @JsonIgnore
+    private Map<String, Object> additionalProperties = new HashMap<String, Object>();
+
+    @JsonProperty("publishers")
```
Comment (jeymarsonwolox) [inline src/main/java/wolox/training/integration/dto/Book.java]: you don't need `JsonProperty` in getters and setters

File: src/main/java/wolox/training/repositories/BookRepository.java [hunk 2]
```diff
@@ -13,4 +14,13 @@
...
      */
     Optional<Book> findByAuthor(String author);
+
+    /**
+     * This method get a book by isbn
+     *
+     * @param isbn
+     * @return
+     */
+    @Query("SELECT b FROM Book b WHERE b.isbn = ?1")
```
Comment (jeymarsonwolox) [inline src/main/java/wolox/training/repositories/BookRepository.java]: query is not necesary, use `Optional<Book> findByIsbn(String isbn);`

File: src/main/java/wolox/training/integration/dto/Author.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({
+    "url",
+    "name"
+})
```
Comment (jeymarsonwolox) [inline src/main/java/wolox/training/integration/dto/Author.java]: `@JsonPropertyOrder` is not necessary but if you want to keep it use constants in the values

File: src/main/java/wolox/training/commons/Constants.java [hunk 4]
```diff
@@ -26,4 +26,20 @@
      * Constant for wrong value
      */
     public static String USER_INCORRECT = "User and id incorrect";
+    /**
+     * Constant for connect with external service
+     */
+    public static String URL = "https://openlibrary.org/api/books?bibkeys=ISBN:{isbn}&format={format}&jscmd={jscmd}";
```
Comment (jeymarsonwolox) [inline src/main/java/wolox/training/commons/Constants.java]: extract this url to  properties file

**Review by jeymarsonwolox — CHANGES_REQUESTED** [review]

File: src/main/java/wolox/training/controllers/BookController.java [hunk 5]
```diff
@@ -91,4 +102,27 @@ public Book findOne(@PathVariable Long id) {
...
+     * This method find a Book by ISBN, if not return find in a external service and persist in data
+     * base
+     *
+     * @param isbn
+     * @return
+     */
+    @GetMapping()
+    public ResponseEntity findByIsbn(@RequestParam("isbn") String isbn) {
+        Book book = bookRepository.findByIsbn(isbn).orElse(null);
+        if (book == null) {
```
Comment (martinremis) [inline src/main/java/wolox/training/controllers/BookController.java]: You could use Optional to handle this without a null comparison.
Something like:
```
Book book = bookRepository.findByIsbn(isbn);
if (book.isPresent()) {
  return book.get();
}
// else
.....
```

File: src/main/java/wolox/training/dto/AuthorDTO.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+public class AuthorDTO {
+
+    public AuthorDTO() {
+    }
+
+    public AuthorDTO(String name) {
+        this.name = name;
+    }
+
+    private String name;
```
Comment (martinremis) [inline src/main/java/wolox/training/dto/AuthorDTO.java]: Place variables before constructors in classes.

File: src/main/java/wolox/training/dto/PublishersDTO.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package wolox.training.dto;
+
+public class PublishersDTO {
+
+    public PublishersDTO() {
+    }
+
+    private String name;
```
Comment (martinremis) [inline src/main/java/wolox/training/dto/PublishersDTO.java]: Same

File: src/main/java/wolox/training/integration/OpenLibraryService.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+    public BookDTO bookInfo(String isbn) throws ResponseStatusException {
+        RestTemplate restTemplate = new RestTemplate();
+        Map<String, String> vars = new HashMap<>();
+        vars.put("isbn", isbn);
+        vars.put("format", Constants.FORMAT);
+        vars.put("jscmd", Constants.TYPE);
+
+        ObjectNode objectNode = restTemplate.getForObject(Constants.URL, ObjectNode.class, vars);
+        objectNode = (ObjectNode) objectNode.get(Constants.ISBN.concat(isbn));
```
Comment (martinremis) [inline src/main/java/wolox/training/integration/OpenLibraryService.java]: There is no need for this line, you could add the `.get()` method to the previous one.

File: src/main/java/wolox/training/mapper/BookMapper.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+import org.mapstruct.Mapping;
+import org.mapstruct.Mappings;
+import org.mapstruct.ReportingPolicy;
+import wolox.training.dto.BookDTO;
+import wolox.training.integration.dto.Book;
+
+@Mapper(unmappedTargetPolicy = ReportingPolicy.IGNORE, componentModel = "spring")
+public interface BookMapper {
+
+    @Mappings({
```
Comment (martinremis) [inline src/main/java/wolox/training/mapper/BookMapper.java:13]: MapStruct has an implicit mapper. If the name of the fields are the same, like in most of this case, the Mapping annotation is not needed.

**Review by martinremis — CHANGES_REQUESTED** [review]

**Review by jeymarsonwolox — APPROVED** [review]

**Review by martinremis — APPROVED** [review]

**Pull request merged**